### PR TITLE
Refactors shared preferences

### LINKED ISSUE #1011: Refactor the use of SharedPreferences
In general, the current approach of using shared preferences is not at all organised. We need a better approach to handle them across the entire codebase.

@lognaturel has some idea regarding this which was mentioned during a conversation in #collect-code which has been pasted below. 
> I haven't deeply considered it yet but it might be something we can represent as enums. Like a `CollectGeneralSettings` type and `CollectAdminSettings`. Then each one can have a key, a default, a corresponding `CollectAdminSettings` in the case of general settings..
`getEnumConstants` might be a way to approach getting a list of all preferences.

### PR DESCRIPTION
**Pull request by shobhitagarwal1612**
Resolves #1011

### CODE REVIEW

**Comment by lognaturel** [comment]
At first pass, this approach looks good! I'm going to need to block off some time to go over it more carefully and consider alternatives. Will do that by feature freeze time.

File: collect_app/src/main/java/org/odk/collect/android/utilities/SharedPreferencesUtils.java [hunk 1]
```diff
@@ -77,205 +51,84 @@ private JSONObject getModifiedPrefs(Collection<String> keys) throws JSONExceptio
...
         ToastUtils.showLongToast(context.getString(R.string.successfully_imported_settings));
     }
 
-    private Collection<String> getAllKeys() {
-        Collection<String> keys = new ArrayList<>();
-        keys.addAll(ALL_GENERAL_KEYS);
-        keys.add(KEY_ADMIN_PW);
+    private Collection<String> getAllGeneralKeys() {
+        Collection<String> keys = GENERAL_KEYS.keySet();
         keys.add(KEY_PASSWORD);
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/utilities/SharedPreferencesUtils.java]: `keySet` doesn't support `add`, see [here](http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html#keySet%28%29). I think the right thing to do is to use the copy constructor.

File: collect_app/src/main/java/org/odk/collect/android/preferences/PreferenceKeys.java [hunk 2]
```diff
@@ -67,36 +69,33 @@
...
-            KEY_COMPLETED_DEFAULT,
-            KEY_MAP_SDK,
-            KEY_MAP_BASEMAP,
-            OSM_BASEMAP_KEY,
-            GOOGLE_MAPS_BASEMAP_DEFAULT,
-            OSM_MAPS_BASEMAP_DEFAULT,
-            KEY_AUTOSEND_WIFI,
-            KEY_AUTOSEND_NETWORK
-    );
+    private static HashMap<String, Object> getHashMap() {
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/preferences/PreferenceKeys.java:72]: Sorry I didn't get back to you sooner --  I do think language should be included. I also notice analytics is missing.

**Review by lognaturel — CHANGES_REQUESTED** [review]
I think this is a big improvement! And it does appear to fix the crash I was getting which is great.

Couple of things that need fixing marked inline.

We should use these classes whenever prefs are accessed. I think I'd prefer `get` in the admin class to also refer to a `HashMap` with explicit defaults like the one in general prefs. But this can happen later -- refactoring is a constant exercise!

**Review by lognaturel — APPROVED** [review]

**Comment by shobhitagarwal1612** [comment]
@lognaturel I didn't use a `hashmap` for admin prefs because the default value for all boolean preferences is `true`

**Comment by lognaturel** [comment]
@shobhitagarwal1612 That's true... for now! I don't love the idea that adding a new admin pref with a non-true default requires also making sure the get method is updated. On the general prefs side, there's only one change to make which is great. But it's not a very big deal and we can certainly leave it as we keep considering how the preferences menu will evolve.

**Comment by shobhitagarwal1612** [comment]
@lognaturel Also with this PR being merged, I think we should refactor the use of shared preferences used elsewhere in the code too. With that we can make the admin prefs consistent too

**Comment by lognaturel** [comment]
@shobhitagarwal1612 Yes, I agree with you.

**Comment by shobhitagarwal1612** [comment]
@lognaturel We should open an issue regarding that to track the status of global refactoring of shared preferences

**Pull request merged**